mne/time_frequency/csd.py

```diff
@@ -106,110 +561,32 @@ def csd_epochs(epochs, mode='multitaper', fmin=0, fmax=np.inf,
-
-    if projs is None:
-        projs = cp.deepcopy(epochs.info['projs'])
+    See Also
+    --------
+    csd_array
+    """
+    if mode == 'fourier':
+        csd = csd_fourier(epochs, fmin=fmin, fmax=fmax, tmin=tmin, tmax=tmax,
+                          n_fft=n_fft, verbose=verbose)
+    elif mode == 'multitaper':
+        csd = csd_multitaper(epochs, fmin=fmin, fmax=fmax, tmin=tmin,
+                             tmax=tmax, n_fft=n_fft, bandwidth=mt_bandwidth,
+                             adaptive=mt_adaptive, low_bias=mt_low_bias,
+                             verbose=verbose)
     else:
-        projs = cp.deepcopy(projs)
-
-    picks_meeg = pick_types(epochs[0].info, meg=True, eeg=True, eog=False,
-                            ref_meg=False, exclude='bads')
-    ch_names = [epochs.ch_names[k] for k in picks_meeg]
-
-    # Preparing time window slice
-    tstart, tend = None, None
-    if tmin is not None:
-        tstart = np.where(epochs.times >= tmin)[0][0]
-    if tmax is not None:
-        tend = np.where(epochs.times <= tmax)[0][-1] + 1
-    tslice = slice(tstart, tend, None)
-    n_times = len(epochs.times[tslice])
-    n_fft = n_times if n_fft is None else n_fft
-
-    # Preparing frequencies of interest
-    sfreq = epochs.info['sfreq']
-    orig_freqs = np.fft.rfftfreq(n_fft, 1. / sfreq)
-    freq_mask = (orig_freqs > fmin) & (orig_freqs < fmax)
-    freqs = orig_freqs[freq_mask]
-    n_freqs = len(freqs)
-
-    if n_freqs == 0:
-        raise ValueError('No discrete fourier transform results within '
-                         'the given frequency window. Please widen either '
-                         'the frequency window or the time window')
-
-    # Preparing for computing CSD
-    logger.info('Computing cross-spectral density from epochs...')
-    window_fun, eigvals, n_tapers, mt_adaptive = _compute_csd_params(
-        n_times, sfreq, mode, mt_bandwidth, mt_low_bias, mt_adaptive)
-
-    csds_mean = np.zeros((len(ch_names), len(ch_names), n_freqs),
-                         dtype=complex)
-
-    # Picking frequencies of interest
-    freq_mask_mt = freq_mask[orig_freqs >= 0]
-
-    # Compute CSD for each epoch
-    n_epochs = 0
-    for epoch in epochs:
-        epoch = epoch[picks_meeg][:, tslice]
-
-        # Calculating Fourier transform using multitaper module
-        csds_epoch = _csd_array(epoch, sfreq, window_fun, eigvals, freq_mask,
-                                freq_mask_mt, n_fft, mode, mt_adaptive)
-
-        # Scaling by number of samples and compensating for loss of power due
-        # to windowing (see section 11.5.2 in Bendat & Piersol).
-        if mode == 'fourier':
-            csds_epoch /= n_times
-            csds_epoch *= 8 / 3.
-
-        # Scaling by sampling frequency for compatibility with Matlab
-        csds_epoch /= sfreq
-
-        csds_mean += csds_epoch
-        n_epochs += 1
-
-    csds_mean /= n_epochs
-
-    logger.info('[done]')
```

Comment:
does your PR change any old result?

Reply:
No. The 'fourier' and 'multitaper' methods produce exactly the same output as before.